orttraining/orttraining/training_ops/cpu/collective/adasum_kernels.cc

```diff
@@ -0,0 +1,72 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "orttraining/training_ops/cpu/collective/adasum_kernels.h"
+#include "orttraining/training_ops/cpu/controlflow/common.h"
+#include "orttraining/core/framework/communication/mpi/mpi_context.h"
+
+namespace onnxruntime {
+namespace contrib {
+
+ONNX_OPERATOR_KERNEL_EX(
+    AdasumAllReduce,
+    kMSDomain,
+    1,
+    kCpuExecutionProvider,
+    KernelDefBuilder()
+        .Alias(AliasRange<0, 0>(0, 1024))
+        .TypeConstraint("T", DataTypeImpl::AllIEEEFloatTensorTypes()),
+    AdasumAllReduce);
+
+Status AdasumAllReduce::Compute(OpKernelContext* context) const {
+
+  // Get tensor count
+  const int num_tensors = context->InputCount();
+  std::vector<int> tensor_element_counts;
+  std::vector<size_t> tensor_offsets;
+  std::vector<size_t> tensor_sizes;
+
+  int64_t total_recv_buffer_len = 0;
+
+  size_t size_in_bytes = 0;
+  for (int i = 0; i < num_tensors; ++i) {
+    const Tensor* x_tensor = context->Input<Tensor>(i);
+    tensor_offsets.push_back(size_in_bytes);
+
+    size_in_bytes = x_tensor->SizeInBytes();
+    total_recv_buffer_len += size_in_bytes;
+
+    tensor_sizes.push_back(size_in_bytes);
+    tensor_element_counts.push_back((int)x_tensor->Shape().Size());
+  }
+  AllocatorPtr allocator;
+  ORT_RETURN_IF_ERROR(context->GetTempSpaceAllocator(&allocator));
+
+  // Allocate fusion buffer.
+  auto data_buffer = allocator->Alloc(total_recv_buffer_len);
```

Comment:
What happens with reporting failures here?  Is that handled inside Alloc?

Reply:
yes, error is checked against different EP's allocator and reported there.